### ISSUE TITLE
render need to be called explicitly for nest child heresy components 

### ISSUE THREAD

**Issue by AhnafCodes** [WebReflection/heresy#22]
When I nest heresy components. 
"render" of the child component is not called after setProps from parent. https://codepen.io/ahnafcodes/pen/QWWoMYE

 i can work around this by calling render explicitly in child which is leads to unnecessary re-render the first time around. 
https://codepen.io/ahnafcodes/pen/gOOExEY
Also at times i  am running into a weird render issues when calling render for nested component inside setState, "Type Error: item is undefined". I need to isolate its recreation for that one. call stack for that : 

![image](https://user-images.githubusercontent.com/7367600/69355546-94fc5e80-0c47-11ea-9cb4-7bf1cc82c01b.png)

**Comment by AhnafCodes** [comment]
oddly one left-hand-side works and right-hand-side fails.
![image](https://user-images.githubusercontent.com/7367600/69378196-7d3acf80-0c73-11ea-86a9-c03de23216ca.png)
failing: https://codepen.io/ahnafcodes/pen/QWWoMYE
working without re-render - https://codepen.io/ahnafcodes/pen/BaaEaQm
May  heresy html`` function doesn't like again logic inside it.

**Comment by WebReflection** [comment]
apologies for my delayed answer, but I'm not having much time these days to dig more, so thanks for all the details, I'll try to dig more ASAP about this

**Comment by AhnafCodes** [comment]
Thanks for all your work,  I greatly appreciate your prompt on this and earlier issue.

 I and may be others are used writing in style which works with React, Angular or Jquery or any other library which might not be the best or right way to do things. When in DOMland I need understand I am in DOMland, I often forget  classList doesn't return an Array and end up using Array methods on it and only to regret later.

**Comment by AhnafCodes** [comment]
![image](https://user-images.githubusercontent.com/7367600/69441735-dc025680-0d10-11ea-81c5-9e46778d30d9.png)
This is another error I am running into when i nesting Heresy components along with regular span's or buttons. Say in the structure below when changing state to render expression-heresy component, which has both span's and my heresy components. Hole for spans get rendered, heresy component are not. If after set props I call this.render explicitly I run into the above issue.   
 <div is="my-div">  
     <div class="controls" >  
         <button>+ <button> 
         <button>  -  <button>
          <button> *<button>
          <button>  []<button>
     </div>
      <div class="values">  
         <button>1 <button> 
         <button>  2  <button>
          <button> 3<button>
          ....
       </div>
   <div is="expression">
        <span class="value">1<span>
        <span class="operation">+</span>
        <span class="value">2<span>
        <span class="operation">*</span>
        <div is="group">
            <span class="value">5<span>
             <span class="operation">-</span>
             <span class="value">2<span>
         </div>
   </div>

**Comment by WebReflection** [comment]
to have an understanding of what's going on, a screenshot of failures in a stack doesn't really help ... a good bug report has a minimal implementation that reveals a bug, as code, or Code Pen, as you did before, and no errors screnshots, as these can't be investigated as images, or stack, without an example easy to reproduce.

Accordingly, if this is another bug, different from the first one, please do open another ticket with a reproducible (minimal) example, so I can fix one bug per time.

Thanks for your understanding.

**Comment by WebReflection** [comment]
Just FYI, I've written a guide on how to file issues https://medium.com/@WebReflection/how-to-file-a-github-lab-issue-7080b8c0cef5 and I hope it'll help you next time you have an issue here, or elsewhere.

**Comment by WebReflection** [comment]
> When I nest heresy components "render" of the child component is not called after setProps from parent. https://codepen.io/ahnafcodes/pen/QWWoMYE

When you receive props, you should render. I don't know why you want to avoid rendering in general, since it's the cheapest operation you have in heresy/lighterhtml, but you think it's a good idea to create every single time props are passed new `setState` and `getState` methods.

```js
    set props(props) {
    this._props = props;
    let State = { color: this.props.color };

    this.setState = (newState, callback) => {
        State = { ...State, ...newState };
        return ((render, callback) => {
            render();
            callback && callback()
        })(() => this.render(),callback)
    };

    this.getState = () => ({ ...State });
    this.render(); // <--- THERE, FIXED
  }

```

If you just pass props but you have setter that doesn't render, heresy can't do much in there, isn't it?

TL;DR when props changes, or whenever it's set, call render, because setting the props is exactly the same as changing the state.

For any other issue, please file an issue a part, as I can't deal with more than one issue per time.

Thanks.

**Comment by WebReflection** [comment]
P.S.

```js
class ControlledPlayground extends HTMLDivElement {
    constructor(...args) {
        super(...args);
```

this is how you should **always** use `super`, or the polyfill for legacy browsers won't work.

**Comment by WebReflection** [comment]
P.S. 2
```js
constructor() {
        super();
    }
```
this is completely unnecessary, redundant, and broken for legacy browsers (no `...args`).

If you don't need to do anything on the `constructor`, just don't define a constructor.

Moreover, heresy gives you an `oninit()` method, which you use, and you should **always** prefer `init()` over `constructor`. If you use contrsuctor and init, you are causing yourself problems with polyfills and the element lifecycle. Just use `oninit()`, which is granted to be called once, like a constructor, and it's granted to give you the node you are after as a context.

Any issue related to the constructor will be ignored in this repository, as it's explained the `constructor` should not be used, as the library grants `oninit()` is the right place to initialize the element, because in the constructor the element is potentially not upgraded yet, and Safari has broken constructors, so everything might fail in there.

**Comment by WebReflection** [comment]
P.S. 3

this is how you render arrays, not by returning empty holes, by returning only what's needed to be rendered.

```js
renderControls() {
        console.log("begin render controls");
        const controlsState = this.getState().controlsState;
        return Object.keys(controlsState)
                      .filter(controlState => controlsState[controlState])
                      .map(controlState => html`<button class=${controlState}>${controlState}</button>`);
    }
```

if you conditionally returns html`` instead of a clean Array with only the stuff you need, you are causing unnecessary issues to the engine, way worse than calling `render()` more than once, 'cause `render`, once again, is your best friend, not something to avoid.

**Comment by AhnafCodes** [comment]
> create every single time

I don't want to do something like "this.state" so the state look like guts-out and I also want to be able derived State from props. Yes I am foolishly creating this.setState and this.getState "every single time". may be i could avoid that by  "short-curcuiting using ||" or even good old "if condition.

**Comment by WebReflection** [comment]
I think both method should be in the class, but this last iteration looks less expensive to me 👍
